### ISSUE TITLE
[Investigation]: Support React 19

### ISSUE THREAD

**Issue by yannbf** [storybookjs/storybook#29805]
### Describe the bug

React 19 is currently at the RC stage. We should investigate all the changes that are needed to support it before release:

- Make sure that React 19 is supported in all packages that have it as direct dependency or peer dependency
  - @storybook/addon-docs
  - @storybook/addon-links
  - @storybook/ember
  - @storybook/blocks
  - @storybook/react
  - @storybook/react-vite
  - @storybook/react-webpack5
  - @storybook/react-dom-shim
  - @storybook/react-webpack
  - @storybook/experimental-nextjs-vite
  - @storybook/react-native-web-vite
  - @storybook/nextjs
  - @storybook/icons
- Make sure React 19 features are tested in sandboxes
- Update [usage of act](https://bsky.app/profile/phry.dev/post/3lchyxhe2ts2b) in places [like this](https://github.com/storybookjs/storybook/blob/ceb8387e24f9b401bd3fc693dbb0f5b220a389d0/code/renderers/react/src/portable-stories.tsx#L66)
- (optional) Update addon-toolkit to use React 19

### Reproduction link

-

### System

```bash
-
```

**Comment by JReinhold** [comment]
I don't think we want to update any direct dependencies, as that would be a breaking change, at least if we did it in `@storybook/addon-docs` or `storybook`.
We also shouldn't upgrade `devDependencies` in addons, **and especially not `addon-kit`**, because the manager will _still_ be using React 18, so the addons' React version is ignored. Meaning if you're building an addon with React 19 APIs, they won't work as soon as that addon is bundled into the manager at runtime.

**Comment by ianzone** [comment]
it's released now

**Comment by silverwind** [comment]
Would appreciate if these `peerDependencies` could be upgrade to support the released version:

https://github.com/storybookjs/storybook/blob/052f6ce95a69ad62006262d980724c236c946fbd/code/renderers/react/package.json#L96-L97

**Comment by yannbf** [comment]
Hey @silverwind they should support the released version. Are you having any problems?

**Comment by silverwind** [comment]
Ah, you are right, it's not an issue. I wasn't aware that prerelease ranges also satisfy release versions. This works with npm for me without getting a `peerDependencies` error:

```json
{
  "dependencies": {
    "@storybook/react": "8.4.7",
    "react": "19.0.0"
  }
}
```

**Comment by kerryj89** [comment]
It seems `addons.TOOLS` and `addons.TOOLEXTRA` do not work with React 19 and cause the following error which probably doesn't mean much to you, but anywhow:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'A')
    at Y (manager-bundle.js:7:726)
    at J (manager-bundle.js:12:268)
    at P.jsx (manager-bundle.js:16:979)
    at render (manager-bundle.js:16:1915)
    at Xh (globals-runtime.js:3339:13)
    at Wk (globals-runtime.js:5796:23)
    at Vk (globals-runtime.js:5549:13)
    at Uk (globals-runtime.js:5541:27)
    at Jk (globals-runtime.js:5530:9)
    at Ok (globals-runtime.js:5306:81)
```

 If I downgrade to React 18.x it works as expected.

Reproduction:

```tsx
// .storybook/manager.tsx

import { addons, types } from '@storybook/manager-api';

addons.register('test', () => {
  addons.add('test', {
    type: types.TOOLEXTRA,
    title: 'test',
    paramKey: 'test',
    render: (props: any) => <>test</>,
  });
});
```

**Comment by terrymun** [comment]
Can confirm that post-React19 upgrade, everything works with the latest version of Storybook (8.4.7) with the exception of our in-house addons: we have encountered the same error as @kerryj89. Temporarily disabling the addon works for now 👍

**Comment by Tottzi** [comment]
Hei Yo,
I still face this issue but only with one addon. 
// 'storybook-addon-apollo-client',
This does not work, only gives the
Uncaught TypeError: Cannot read properties of undefined (reading 'A')
    at Y (manager-bundle.js:7:726)
    at J (manager-bundle.js:12:268)
    at P.jsx (manager-bundle.js:16:979)
    at render (manager-bundle.js:16:1915)
    at Xh (globals-runtime.js:3339:13)
    at Wk (globals-runtime.js:5796:23)
    at Vk (globals-runtime.js:5549:13)
    at Uk (globals-runtime.js:5541:27)
    at Jk (globals-runtime.js:5530:9)
    at Ok (globals-runtime.js:5306:81)

**Comment by mrginglymus** [comment]
There are a number of bundled dependencies in storybook that have incompatible types (mostly due to still using the global `JSX` namespace); I've started a PR to fix the ones I've spotted so far here: #30031

**Comment by Sidnioulz** [comment]
Thanks for the quick fix @yannbf!

Do you know if this is being backported to 8.4? Or will it be in 8.5+?

**Comment by gvillo** [comment]
+1

**Comment by yannbf** [comment]
> Thanks for the quick fix [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1671563?s=64&amp;u=292df89092fe8845fc79d859b11802e6f01b4be1&amp;v=4">@yannbf](https://github.com/yannbf?rgh-link-date=2024-12-12T08%3A24%3A15.000Z)!
> 
> Do you know if this is being backported to 8.4? Or will it be in 8.5+?

Most of the work we are doing is towards 8.5, which is currently in beta so should be released soon!

**Comment by savanesoff** [comment]
> > Thanks for the quick fix [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1671563?s=64&amp;u=292df89092fe8845fc79d859b11802e6f01b4be1&amp;v=4">@yannbf](https://github.com/yannbf?rgh-link-date=2024-12-12T08%3A24%3A15.000Z)!
> > Do you know if this is being backported to 8.4? Or will it be in 8.5+?
> 
> Most of the work we are doing is towards 8.5, which is currently in beta so should be released soon!

When can we expect this to be avaialble?
This is a magor issue as a lot depends on automation.
Is there a way to run a beta instad?

**Comment by terrymun** [comment]
> > > Thanks for the quick fix [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1671563?s=64&amp;u=292df89092fe8845fc79d859b11802e6f01b4be1&amp;v=4">@yannbf](https://github.com/yannbf?rgh-link-date=2024-12-12T08%3A24%3A15.000Z)!
> > > Do you know if this is being backported to 8.4? Or will it be in 8.5+?
> > 
> > 
> > Most of the work we are doing is towards 8.5, which is currently in beta so should be released soon!
> 
> When can we expect this to be avaialble? This is a magor issue as a lot depends on automation. Is there a way to run a beta instad?

If you're in a pinch you can always give the latest beta release for 8.5.0 a try: https://github.com/storybookjs/storybook/releases/tag/v8.5.0-beta.5

**Comment by savanesoff** [comment]
> > > > Thanks for the quick fix [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1671563?s=64&amp;u=292df89092fe8845fc79d859b11802e6f01b4be1&amp;v=4">@yannbf](https://github.com/yannbf?rgh-link-date=2024-12-12T08%3A24%3A15.000Z)!
> > > > Do you know if this is being backported to 8.4? Or will it be in 8.5+?
> > > 
> > > 
> > > Most of the work we are doing is towards 8.5, which is currently in beta so should be released soon!
> > 
> > 
> > When can we expect this to be avaialble? This is a magor issue as a lot depends on automation. Is there a way to run a beta instad?
> 
> If you're in a pinch you can always give the latest beta release for 8.5.0 a try: https://github.com/storybookjs/storybook/releases/tag/v8.5.0-beta.5

Thanks for speedy reply @terrymun 
I've never has the opportunity to install/use unrelease to npm registry packages, any pointers on how I go about using beta?
Thanks again!

**Comment by savanesoff** [comment]
> [@savanesoff](https://github.com/savanesoff) You install it just like how you'd [install a specific version](https://docs.npmjs.com/cli/v9/commands/npm-install) of any dependency, e.g. if you're npm then it's just `npm install storybook@8.5.0-beta.5`

Oh my... LOL. Thanks! 😄 
I was searching for it in NPM registry and coulnd't find it, untill I really looked for it 😇 My bad...
You're a saviour!

**Comment by savanesoff** [comment]
> [@savanesoff](https://github.com/savanesoff) 8.5.0-beta.5 has been released: check the [releases page](https://github.com/storybookjs/storybook/releases) or [npmjs](https://www.npmjs.com/package/storybook?activeTab=versions). You install it just like how you'd [install a specific version](https://docs.npmjs.com/cli/v9/commands/npm-install) of any dependency, e.g. if you're npm then it's just `npm install storybook@8.5.0-beta.5`

I'm running into issue with one of my projects that uses `@storybook/addon-postcss` addon while using `8.5.0-beta`
There doesn't appear to be an update on the addon, so I'm using `^v2`

Another one of my projects doesn't use this addon and it works perfectly fine.

There are no errors to report, neither in browser console or terminal. The storybook UI just hangs while trying to load a story. 

I get this warning in my terminal:

```
The following packages are incompatible with Storybook 8.5.0-beta.5 as they depend on different major versions of Storybook packages:
WARN - @storybook/addon-postcss@2.0.0
```

but that's about it.

Could you please advise on what I could do to mitigate this issue?

Thank you!

**Comment by rrogowski** [comment]
Hi there! Will React 19 be supported for Storybook 7? Or is an upgrade to Storybook 8 required?

**Comment by yannbf** [comment]
> Hi there! Will React 19 be supported for Storybook 7? Or is an upgrade to Storybook 8 required?

Hey there, there are no plans to backport this to Storybook 7. Please upgrade to Storybook 8 🙏

**Comment by Kunkka0822** [comment]
Hello @yannbf! Happy New Year!

I'm getting the error while working with React 19 and Storybook 8.5.0-beta.5.

```
Error
A React Element from an older version of React was rendered. This is not supported. It can happen if:
- Multiple copies of the "react" package is used.
- A library pre-bundled an old copy of "react" or "react/jsx-runtime".
- A compiler tries to "inline" JSX instead of using the runtime.
Call Stack
 throwOnInvalidObjectType
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:4476:15
 reconcileChildFibersImpl
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:5410:11
 undefined
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:5449:33
 reconcileChildren
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:7772:13
 updateFunctionComponent
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:8069:7
 beginWork
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:9717:18
 runWithFiberInDEV
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:577:16
 performUnitOfWork
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:15078:22
 workLoopSync
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:14904:41
 renderRootSync
  vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:14884:11
```

I've noticed the error comes from alias importing with Storybook, but I am not sure how to resolve it.
This is my main.ts.

```
import type { StorybookConfig } from '@storybook/react-webpack5';
import path from 'path';

const config: StorybookConfig = {
  stories: ['../src/**/*.mdx', '../src/**/*.stories.@(js|jsx|mjs|ts|tsx)'],
  addons: ['@storybook/preset-create-react-app', '@storybook/addon-links', '@storybook/addon-essentials'],
  framework: {
    name: '@storybook/react-webpack5',
    options: {},
  },
  staticDirs: ['..\\public'],
  webpackFinal: async (config) => {
    config.resolve = {
      ...config.resolve,
      alias: {
        '@': path.resolve(__dirname, '..', 'src'),
      },
    };

    return config;
  },
};
export default config;
```

Here webpack.config.js
```
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin');
const ESLintPlugin = require('eslint-webpack-plugin');

module.exports = {
  mode: 'development',
  entry: './src/index.tsx',
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'bundle.js',
  },
  devServer: {
    static: {
      directory: path.resolve(__dirname, 'dist'),
    },
    port: 3000,
    open: true,
    hot: true,
    compress: true,
    historyApiFallback: true,
  },
  module: {
    rules: [
      {
        test: /\.css$/i,
        include: path.resolve(__dirname, 'src'),
        use: ['style-loader', 'css-loader', 'postcss-loader'],
      },
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: /node_modules/,
      },
      {
        test: /\.(png|jpe?g|gif|jp2|webp)$/,
        loader: 'file-loader',
        options: {
          name: '[name].[ext]',
        },
      },
      {
        test: /\.svg$/i,
        issuer: /\.[jt]sx?$/,
        use: ['@svgr/webpack', 'url-loader'],
      },
    ],
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: './src/index.html',
    }),
    new ESLintPlugin({
      extensions: ['ts'],
    }),
  ],
  resolve: {
    extensions: ['.tsx', '.ts', '.js'],
    plugins: [
      new TsconfigPathsPlugin({
        extensions: ['.js', '.jsx', '.json', '.ts', '.tsx'],
      }),
    ],
  },
};
```

**Comment by Kunkka0822** [comment]
> Hello [@yannbf](https://github.com/yannbf)! Happy New Year!
> 
> I'm getting the error while working with React 19 and Storybook 8.5.0-beta.5.
> 
> ```
> Error
> A React Element from an older version of React was rendered. This is not supported. It can happen if:
> - Multiple copies of the "react" package is used.
> - A library pre-bundled an old copy of "react" or "react/jsx-runtime".
> - A compiler tries to "inline" JSX instead of using the runtime.
> Call Stack
>  throwOnInvalidObjectType
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:4476:15
>  reconcileChildFibersImpl
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:5410:11
>  undefined
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:5449:33
>  reconcileChildren
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:7772:13
>  updateFunctionComponent
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:8069:7
>  beginWork
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:9717:18
>  runWithFiberInDEV
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:577:16
>  performUnitOfWork
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:15078:22
>  workLoopSync
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:14904:41
>  renderRootSync
>   vendors-node_modules_storybook_react-dom-shim_dist_react-18_mjs.iframe.bundle.js:14884:11
> ```
> 
> I've noticed the error comes from alias importing with Storybook, but I am not sure how to resolve it. This is my main.ts.
> 
> ```
> import type { StorybookConfig } from '@storybook/react-webpack5';
> import path from 'path';
> 
> const config: StorybookConfig = {
>   stories: ['../src/**/*.mdx', '../src/**/*.stories.@(js|jsx|mjs|ts|tsx)'],
>   addons: ['@storybook/preset-create-react-app', '@storybook/addon-links', '@storybook/addon-essentials'],
>   framework: {
>     name: '@storybook/react-webpack5',
>     options: {},
>   },
>   staticDirs: ['..\\public'],
>   webpackFinal: async (config) => {
>     config.resolve = {
>       ...config.resolve,
>       alias: {
>         '@': path.resolve(__dirname, '..', 'src'),     //// <-- error from this line
>       },
>     };
> 
>     return config;
>   },
> };
> export default config;
> ```
> 
> Here webpack.config.js
> 
> ```
> const path = require('path');
> const HtmlWebpackPlugin = require('html-webpack-plugin');
> const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin');
> const ESLintPlugin = require('eslint-webpack-plugin');
> 
> module.exports = {
>   mode: 'development',
>   entry: './src/index.tsx',
>   output: {
>     path: path.resolve(__dirname, 'dist'),
>     filename: 'bundle.js',
>   },
>   devServer: {
>     static: {
>       directory: path.resolve(__dirname, 'dist'),
>     },
>     port: 3000,
>     open: true,
>     hot: true,
>     compress: true,
>     historyApiFallback: true,
>   },
>   module: {
>     rules: [
>       {
>         test: /\.css$/i,
>         include: path.resolve(__dirname, 'src'),
>         use: ['style-loader', 'css-loader', 'postcss-loader'],
>       },
>       {
>         test: /\.tsx?$/,
>         use: 'ts-loader',
>         exclude: /node_modules/,
>       },
>       {
>         test: /\.(png|jpe?g|gif|jp2|webp)$/,
>         loader: 'file-loader',
>         options: {
>           name: '[name].[ext]',
>         },
>       },
>       {
>         test: /\.svg$/i,
>         issuer: /\.[jt]sx?$/,
>         use: ['@svgr/webpack', 'url-loader'],
>       },
>     ],
>   },
>   plugins: [
>     new HtmlWebpackPlugin({
>       template: './src/index.html',
>     }),
>     new ESLintPlugin({
>       extensions: ['ts'],
>     }),
>   ],
>   resolve: {
>     extensions: ['.tsx', '.ts', '.js'],
>     plugins: [
>       new TsconfigPathsPlugin({
>         extensions: ['.js', '.jsx', '.json', '.ts', '.tsx'],
>       }),
>     ],
>   },
> };
> ```

**Comment by zeppelinnn** [comment]
Hi @yannbf, I spent several hours yesterday trying to get storybook (8.4.7) working with React 19 (w/ tailwind) but kept running into the following error when booting storybook. I am using `@storybook/experimental-nextjs-vite` as the framework in a monorepo created through turborepo:

```
TypeError: Cannot read properties of undefined (reading 'A')
    at getOwner (http://localhost:6006/node_modules/.cache/storybook/bd45b70ddf2f9c95d1ea5d1e495c39b571d64dc099b1becb6c02aa1a45de3ae3/sb-vite/deps/react_jsx-dev-runtime.js?v=c6a66091:274:47)
    at jsxDEVImpl (http://localhost:6006/node_modules/.cache/storybook/bd45b70ddf2f9c95d1ea5d1e495c39b571d64dc099b1becb6c02aa1a45de3ae3/sb-vite/deps/react_jsx-dev-runtime.js?v=c6a66091:385:59)
    at exports.jsxDEV (http://localhost:6006/node_modules/.cache/storybook/bd45b70ddf2f9c95d1ea5d1e495c39b571d64dc099b1becb6c02aa1a45de3ae3/sb-vite/deps/react_jsx-dev-runtime.js?v=c6a66091:440:16)
    at decorators (http://localhost:6006/src/card.stories.tsx:13:42)
    at hookified (http://localhost:6006/sb-preview/runtime.js:3640:13)
    at http://localhost:6006/sb-preview/runtime.js:4920:17
    at http://localhost:6006/sb-preview/runtime.js:4945:8
    at withOutline (http://localhost:6006/node_modules/.cache/storybook/bd45b70ddf2f9c95d1ea5d1e495c39b571d64dc099b1becb6c02aa1a45de3ae3/sb-vite/deps/@storybook_addon-essentials_outline_preview.js?v=c295f91c:438:43)
    at hookified (http://localhost:6006/sb-preview/runtime.js:3640:13)
    at http://localhost:6006/sb-preview/runtime.js:4920:17
```

I have recreated a minimal reproducible example at https://github.com/zeppelinnn/react-19-storybook-nextjs-vite
After cloning, run `yarn` and navigate to `packages/ui` and run `yarn storybook` to reproduce the issue.

Let me know if you need any other info. Your help is much appreciated

**Comment by Boonstra** [comment]
@zeppelinnn As mentioned by Terry, you can give the Storybook 8.5.0 beta a try to resolve that issue. For me the error is resolved with `storybook@8.5.0-beta.8`.

**Comment by zeppelinnn** [comment]
> [@zeppelinnn](https://github.com/zeppelinnn) As mentioned by Terry, you can give the Storybook 8.5.0 beta a try to resolve that issue. For me the error is resolved with `storybook@8.5.0-beta.8`.

Hey @Boonstra, thanks for the reply! I added a new commit to bump dependencies to the beta version you mentioned (in the example repo I shared) but i'm still seeing the same error. I believe this is specific to the `@storybook/experimental-nextjs-vite` framework - if i try to use `@storybook/nextjs` framework it works fine.

Could you please check again? Thank you!

**Comment by Boonstra** [comment]
> Hey [@Boonstra](https://github.com/Boonstra), thanks for the reply! I added a new commit to bump dependencies to the beta version you mentioned (in the example repo I shared) but i'm still seeing the same error. I believe this is specific to the `@storybook/experimental-nextjs-vite` framework - if i try to use `@storybook/nextjs` framework it works fine.
>
> Could you please check again? Thank you!

Right, I also can't get the Storybook in your repository to run without that error coming up.

**Comment by akaltar** [comment]
I'm trying to upgrade a project to React 19 with storybook, and I've ran into a typescript issue, I've created a reproduction of the issue here: https://stackblitz.com/edit/github-bmd2twki?file=src%2Fstories%2FPage.stories.tsx&preset=node

The error is 
```
'Story' cannot be used as a JSX component.
  Its type 'AnnotatedStoryFn<ReactRenderer, Args>' is not a valid JSX element type.
    Type 'AnnotatedStoryFn<ReactRenderer, Args>' is not assignable to type '(props: any) => ReactNode'.
      Target signature provides too few arguments. Expected 2 or more, but got 1.ts(2786)
```

And it happens when trying to declare JSX decorators like this:

```ts
export const withFullHeight = (Story: StoryFn) => {
  return (
    <div className="h-full">
      <Story />
    </div>
  );
};

```

Is there a different type instead of `StoryFn` I can use to type these decorators that still works with React 19?


PS: Thanks for adding support for React 19 so quickly.

**Comment by terrymun** [comment]
@akaltar I personally prefer typing the function instead of the parameters, since it requires knowledge of the underlying rendered. This should probably work for you:

```tsx
import type { ReactRenderer } from '@storybook/react';
import type { DecoratorFunction } from '@storybook/types';

export const withFullHeight: DecoratorFunction<ReactRenderer> = (Story) => {
  return (
    <div className="h-full">
      <Story />
    </div>
  );
};
```

**Comment by akaltar** [comment]
Thanks, that resolves the problem!

**Comment by terrymun** [comment]
@yannbf Can confirm that with the latest 8.5.0 release, the addon issues mentioned by @kerryj89 (https://github.com/storybookjs/storybook/issues/29805#issuecomment-2526402213) has been fixed.

Thank you so much!

**Comment by mpelham** [comment]
Howdy folks, I have run into a few issues with Addon-a11y 8.50 and React 19:
```
 FAIL  src/stories/Components/Foo/Foo.stories.tsx
  ● Foo/Foo › Default › smoke-test

    page.evaluate: Error: Axe is already running. Use `await axe.run()` to wait for the previous run to finish before starting a new run.
```

I will also note that `addon-ally` will fail when running `npx storybook@next upgrade`.
```
npm ERR! Found: @storybook/addon-a11y@8.5.0
npm ERR! node_modules/@storybook/addon-a11y
npm ERR!   dev @storybook/addon-a11y@"^8.6.0-alpha.0" from the root project
```

**Comment by JReinhold** [comment]
@mpelham could you open a new issue? I'm not sure this is related to React 19

**Comment by Mearman** [comment]
It seems to be behaving better for me. this may be a red herring, but it appears that my addon only works if the `manager` file is `.tsx` even if there is no TSX in it

**Comment by zeppelinnn** [comment]
> > Hey [@Boonstra](https://github.com/Boonstra), thanks for the reply! I added a new commit to bump dependencies to the beta version you mentioned (in the example repo I shared) but i'm still seeing the same error. I believe this is specific to the `@storybook/experimental-nextjs-vite` framework - if i try to use `@storybook/nextjs` framework it works fine.
> > Could you please check again? Thank you!
> 
> Right, I also can't get the Storybook in your repository to run without that error coming up.

I've tried to upgrade to storybook@8.5 but still experiencing the same issue (original [report](https://github.com/storybookjs/storybook/issues/29805#issuecomment-2578125868))

We're blocked on upgrading to react 19 due to this error - any update on progress or an ETA would be greatly appreciated!

**Comment by JReinhold** [comment]
> I've tried to upgrade to storybook@8.5 but still experiencing the same issue (original [report](https://github.com/storybookjs/storybook/issues/29805#issuecomment-2578125868))
> 
> We're blocked on upgrading to react 19 due to this error - any update on progress or an ETA would be greatly appreciated!

@zeppelinnn the reason you're experiencing this issue, is because you're still using Next.js v14 which doesn't support React 19.

Storybook's Next.js integration (`@storybook/nextjs` and `@storybook/experimental-nextjs-vite`) uses the React version that Next.js comes _precompiled with_, and that version must match what you define in Storybook. In your `apps/docs` and and `apps/web` you have `"next": "^14.2.3"`, which only supports React 18. While you also have `"react": "19.0.0"` there, that doesn't actually make a difference because Next.js ships its own version of React, so you're lying to yourself. You can see this if you do something like this in your Next.js application today:

```js
'use client';
import React from 'react';

function SomeClientComponent() {
  React.useEffect(() => {
    console.log('React version:', React.version);
  }, [])
  return null;
}

```

You'll get `React version: 18.3.0-canary-178c267a4e-20241218` in your console.

There are two ways to solve this:

1. Install `next@^15` as a `devDependency` in your `ui` package, and Storybook will use that correctly then. Then the UI package will not depend on your `next` package being hoisted from `apps` and `docs`. Though your UI will now support Next.js 15 and React 19, your app still won't, so you could run into trouble there.

As a side note, you should also install `react` and `react-dom` as `devDependencies` to your `ui` package. Even though they are currently listed as `peerDependencies`, your UI package is not self-contained, which is a common footgun. It only really works right now because Yarn hoists React from your other packages, but that can easily change in the future. If dependencies changes in your _other_ packages, the UI package will break, which is not something you want. It's best practice to list explicit versions of any peer dependencies you have as dev dependencies.

2. Upgrade to Next.js v15 in your apps and docs, so they actually come shipped with React 19.

In reality you probably want to do both.

**Comment by Demianeen** [comment]
Did anyone managed to fix type issue with StoryFn type with function declaration? I mean this error:
```
     ├╴  'StoryComponent' cannot be used as a JSX component.
     │      Its type 'AnnotatedStoryFn<ReactRenderer, ReactRenderer>' is not a valid JSX element type.
     │        Type 'AnnotatedStoryFn<ReactRenderer, ReactRenderer>' is not assignable to type '(props: any) => ReactNode | Promise<ReactNode>'.
     │          Target signature provides too few arguments. Expected 2 or more, but got 1. ts (2786) [7, 5]
```

It works with arrow function expression with `DecoratorFunction<ReactRenderer>` but I don't think we can assign type like this to function declaration.

**Comment by mrginglymus** [comment]
Might be fixed by #30031?

**Comment by akibarika** [comment]
> Howdy folks, I have run into a few issues with Addon-a11y 8.5.0 and React 19:
> 
> ```
>  FAIL  src/stories/Components/Foo/Foo.stories.tsx
>   ● Foo/Foo › Default › smoke-test
> 
>     page.evaluate: Error: Axe is already running. Use `await axe.run()` to wait for the previous run to finish before starting a new run.
> ```
> 
> I will also note that `addon-ally` will fail when running `npx storybook@next upgrade`.
> 
> ```
> npm ERR! Found: @storybook/addon-a11y@8.5.0
> npm ERR! node_modules/@storybook/addon-a11y
> npm ERR!   dev @storybook/addon-a11y@"^8.6.0-alpha.0" from the root project
> ```
> 
> Thanks so much for all the work, and have a great day! I can open up an issue on this if required!

I had a same issue, and I don't think it relates with React 19. I think it's the issue with upgrading to 8.5.0

**Comment by Tottzi** [comment]
For me a Next upgrade to v15.1.5 caused that a11y addon fail. It complains for the same Axe thing
Error: Axe is already running. Use `await axe.run()` to wait for the previous run to finish before starting a new run.

**Comment by Flambe** [comment]
The `Axe is already running` issue isn't related to React (unless it's related to the storybook UI itself). My app is built in Vue and I've got the same issue after updating to 8.5.*

**Comment by valentinpalkovic** [comment]
@Flambe please open a separate issue. Ideally with an reproduction.

**Comment by zeppelinnn** [comment]
> > I've tried to upgrade to storybook@8.5 but still experiencing the same issue (original [report](https://github.com/storybookjs/storybook/issues/29805#issuecomment-2578125868))
> > We're blocked on upgrading to react 19 due to this error - any update on progress or an ETA would be greatly appreciated!
> 
> [@zeppelinnn](https://github.com/zeppelinnn) the reason you're experiencing this issue, is because you're still using Next.js v14 which doesn't support React 19.
> 
> Storybook's Next.js integration (`@storybook/nextjs` and `@storybook/experimental-nextjs-vite`) uses the React version that Next.js comes _precompiled with_, and that version must match what you define in Storybook. In your `apps/docs` and and `apps/web` you have `"next": "^14.2.3"`, which only supports React 18. While you also have `"react": "19.0.0"` there, that doesn't actually make a difference because Next.js ships its own version of React, so you're lying to yourself. You can see this if you do something like this in your Next.js application today:
> 
> 'use client';
> import React from 'react';
> 
> function SomeClientComponent() {
>   React.useEffect(() => {
>     console.log('React version:', React.version);
>   }, [])
>   return null;
> }
> You'll get `React version: 18.3.0-canary-178c267a4e-20241218` in your console.
> 
> There are two ways to solve this:
> 
> 1. Install `next@^15` as a `devDependency` in your `ui` package, and Storybook will use that correctly then. Then the UI package will not depend on your `next` package being hoisted from `apps` and `docs`. Though your UI will now support Next.js 15 and React 19, your app still won't, so you could run into trouble there.
> 
> You can see I did this here which made things work again: [JReinhold/react-19-storybook-nextjs-vite@19a2a92](https://github.com/JReinhold/react-19-storybook-nextjs-vite/commit/19a2a92938a95a9dcdde8d67bab0203431d51e12)
> 
> As a side note, you should also install `react` and `react-dom` as `devDependencies` to your `ui` package. Even though they are currently listed as `peerDependencies`, your UI package is not self-contained, which is a common footgun. It only really works right now because Yarn hoists React from your other packages, but that can easily change in the future. If dependencies changes in your _other_ packages, the UI package will break, which is not something you want. It's best practice to list explicit versions of any peer dependencies you have as dev dependencies.
> 
> 2. Upgrade to Next.js v15 in your apps and docs, so they actually come shipped with React 19.
> 
> In reality you probably want to do both.

that worked, thank you for the help!

**Comment by akibarika** [comment]
I have also posted a discussion (https://github.com/storybookjs/storybook/discussions/30351) with my fix. The root cause of mine is multi-worker

**Comment by terrymun** [comment]
@Eyad-A Storybook 7 has been released quite awhile ago and I do not expect it to be updated to support React 19. You need to upgrade to Storybook 8.5+

**Comment by valentinpalkovic** [comment]
That’s correct. We only patch back critical security fixes to the previous major version.

**Comment by mathiaslaramee** [comment]
Is it in the plans to also upgrade  https://github.com/storybookjs/storybook/blob/052f6ce95a69ad62006262d980724c236c946fbd/code/addons/docs/package.json#L104-L105 ?

**Comment by valentinpalkovic** [comment]
Yes! It seems to be forgotten. Do you want to open a PR?

**Comment by maddyahmad** [comment]
Hi, I have updated react to v19, and after that I am getting an error. I have also updated to v8.5.2, but nothing works.

My component is
import { SnackbarProvider } from "@/store/SnackbarContext";
import { StoryFn } from "@storybook/react";
import { render } from "@testing-library/react";
import { merge } from "lodash";
import React, { createContext, useContext } from "react";
import { Theme, defaultThemeConstants } from '../utils/theme';

interface ThemeProviderProps {
    theme?: Theme
    children?: React.ReactNode
}

const defaultTheme: Theme = { ...defaultThemeConstants }

const ThemeContext = createContext<Theme | undefined>(undefined)

export const ThemeProvider: React.FC<ThemeProviderProps> = (props) => {
    return (
        <ThemeContext.Provider value={merge(defaultTheme, props.theme)}>
            {props.children}
        </ThemeContext.Provider>
    )
}

export const withThemeProvider = (Story: StoryFn) => (
    <ThemeProvider>
        <SnackbarProvider>
            <Story />
        </SnackbarProvider>
    </ThemeProvider>
);


After upgrading react to v19, I am getting an error in withThemeProvider 
'Story' cannot be used as a JSX component.
  Its type 'AnnotatedStoryFn<ReactRenderer, Args>' is not a valid JSX element type.
    Type 'AnnotatedStoryFn<ReactRenderer, Args>' is not assignable to type '(props: any) => ReactNode'.
      Target signature provides too few arguments. Expected 2 or more, but got 1.ts(2786)
(parameter) Story: AnnotatedStoryFn<ReactRenderer, Args>.

For you reference, this is an example of how I am using it
import { withThemeProvider } from '@/theme/ThemeProvider';
import { Meta, StoryObj } from '@storybook/react';
import AccidentCard from './AccidentCard';
import { AccidentCardDoc } from './AccidentCard.doc';

const meta: Meta<typeof AccidentCard> = {
    title: 'Stories/Minor/AccidentCard',
    component: AccidentCard,
    parameters: {
        layout: 'centered',
        actions: { argTypesRegex: '^on.*' },
        docs: {
            description: {
                component: AccidentCardDoc
            }
        }
    },
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof AccidentCard>;


export const Usage: Story = {
    args: {
        iconBackgroundColor: '#FFF1E4',
        CarBrand: 'Toyota Camry',
        CarNumber: '6565 ABC',
        PolicyCardTitle: 'Policy No. 2346576898',
        PolicyCardDescription: 'Aviation Australia College',
        onClick: () => { }
    },
    render: (args) => (
        <div style={{ width: '25rem', boxSizing: 'border-box' }}>
            <AccidentCard {...args}
            />
        </div>
    ),
    decorators: [withThemeProvider],
};

**Comment by mathiaslaramee** [comment]
> That’s correct. We only patch back critical security fixes to the previous major version.

I'm trying but struggling a bit since this is a dependencies update and not just a peerDependencies update. Giving me some troubles when trying to install the dependencies after I specify the new react version.

@JReinhold also mentioned in his first reply that changing this specifically in @storybook/addon-docs is a breaking change.
Any thoughts? @valentinpalkovic

**Comment by mrginglymus** [comment]
@maddyahmad `8.6.0-alpha.3` might fix those problems; that contains a lot of type fixes.

**Comment by mellm0** [comment]
Hi!

It seems we are still getting this issue with the latest Storybook. From debugging, it seems to be when we import a component from an external library inside the addon.

// Custom addon

```tsx
import { LiveProvider } from 'react-live';
 <LiveProvider code={state.code} theme={theme}></LiveProvider>
```

The error I am seeing is similar to the ones posted previously

```
globals-runtime.js:2661 Uncaught TypeError: Cannot read properties of undefined (reading 'A')
    at p (manager-bundle.js:7:744)
    at k (manager-bundle.js:12:280)
    at Co.jsx (manager-bundle.js:16:980)
    at y2 (manager-bundle.js:160:1924)
    at Xh (globals-runtime.js:3339:13)
    at Wk (globals-runtime.js:5796:23)
    at Vk (globals-runtime.js:5549:13)
    at Uk (globals-runtime.js:5541:27)
    at Jk (globals-runtime.js:5530:9)
    at Ok (globals-runtime.js:5306:81)
```

Unfortunately I am not sure how to un-minify the manager-bundle.js to investigate more.

**Comment by shilman** [comment]
Closing and locking this issue since it has become a bit of a dumping ground.

If you have a _specific_ React 19 issue, please file it separately [with a reproduction](https://storybook.new) and we'll be happy to take a look!
